lib/Dialect/TritonGPU/Transforms/Utility.cpp

```diff
@@ -319,28 +322,83 @@ inferSrcEncoding(triton::ExperimentalInterleaveOp op, Attribute encoding) {
       enc.getOrder(), enc.getCTALayout());
 }
 
+static std::optional<Attribute>
+inferTransOpDstEncoding(Attribute srcEnc, ArrayRef<int32_t> order) {
+  // Simply forward to the existing inferTransOpEncoding function.
+  Attribute retEncoding;
+  if (succeeded(
+          srcEnc.getDialect()
+              .getRegisteredInterface<triton::DialectInferLayoutInterface>()
+              ->inferTransOpEncoding(srcEnc, order, retEncoding))) {
+    return retEncoding;
+  }
+  return std::nullopt;
+}
+
+static std::optional<Attribute> inferDstEncoding(triton::TransOp op,
+                                                 Attribute encoding) {
+  return inferTransOpDstEncoding(encoding, op.getOrder());
+}
+
+static std::optional<Attribute> inferSrcEncoding(triton::TransOp op,
+                                                 Attribute encoding) {
+  // We want to solve for srcEnc in
+  //   transpose(srcEnc, order) -> dstEnc.
+  // Given the identity
+  //   transpose(transpose(x, order), inverse(order)) == x,
+  // we can see this is equivalent to
+  //   transpose(dstEnc, inverse(order)) -> srcEnc.
+  return inferTransOpDstEncoding(encoding,
+                                 triton::inversePermutation(op.getOrder()));
+}
+
 std::optional<Attribute> inferSrcEncoding(Operation *op, Attribute encoding) {
+  if (op->hasTrait<mlir::OpTrait::SameOperandsAndResultEncoding>() ||
+      op->hasTrait<mlir::OpTrait::SameLoadStoreOperandsAndResultEncoding>() ||
+      op->hasTrait<mlir::OpTrait::Elementwise>() ||
+      isa<scf::WhileOp, scf::YieldOp, scf::ConditionOp>(op)) {
+    return encoding;
+  }
+
   if (auto reduceOp = dyn_cast<triton::ReduceOp>(op))
     return inferSrcEncoding(reduceOp, encoding);
   if (auto expand = dyn_cast<triton::ExpandDimsOp>(op))
     return inferSrcEncoding(expand, encoding);
   if (auto interleave = dyn_cast<triton::ExperimentalInterleaveOp>(op))
     return inferSrcEncoding(interleave, encoding);
-  if (isa<triton::ReshapeOp, triton::CatOp>(op))
-    return std::nullopt;
-  return encoding;
+  if (auto trans = dyn_cast<triton::TransOp>(op))
+    return inferSrcEncoding(trans, encoding);
+
+  // We have to return `encoding` for ConvertLayoutOp or we get spurrious
+  // conversion ops in the final compiled result.  I think this is due to a
+  // quirk in the impl of RemoveLayoutConversions.  I suppose ConvertLayoutOp
+  // *can* accept the same src and dst encodings, though why would you want to?
+  if (isa<triton::gpu::ConvertLayoutOp>(op))
+    return encoding;
+
```

Comment:
Yup, it's needed.  Otherwise RemoveLayoutConversions does not know what to do with ConvertLayoutOps.

I'm sure RemoveLayoutConversions could be improved though.

Reply:
that’s surprising as I’m pretty sure it was working before. I think this is going to break some of the existing logic of removeLayoutConversion as it usually try to propagate until a convert and stop there. Otherwise we are going to continue walking up the slice incorrectly. I’m surprised this doesn’t break the lit test, I think it’s because we check for the convert layout first in the backward slice logic. If you have an example where it breaks handy I’d be interested in taking a look.

Reply:
oh the problem was that before we were blindly propagating? I guess let’s keep it like that then and I can fix it in a different PR. Would any of the existing lit test handle it?

Reply:
Yes, before we were blindly propagating, and this keeps the old behavior.

If I comment this out in both the inferSrc and inferDst cases, I get failures in TritonGPU/combine.mlir that look real to me.

Reply:
Makes sense, thanks